src/dftbp/dftb/multiexpan.F90

```diff
@@ -0,0 +1,1742 @@
+    nQuadrupole = dimQuadrupole
+  end subroutine getMultiExpanInfo
+
+
+  !> Returns the equivalence to get the correct mixing of charge dependent contributions
+  subroutine getOrbitalEquiv(this, equivDip, equivQuad)
+
+    !> Data structure
+    class(TDftbMultiExpan), intent(inout) :: this
+
+    !> The equivalence vector for cumulative atomic dipole populations
+    integer, intent(out) :: equivDip(:,:)
+
+    !> The equivalence vector for cumulative atomic quadrupole populations
+    integer, intent(out) :: equivQuad(:,:)
+
+    integer :: nAtom, iCount, iSpin, nSpin, iAt, iSp, ii
+
+    nAtom = size(equivDip, dim=2)
+    nSpin = 1
+
+    equivDip(:,:) = 0
+    equivQuad(:,:) = 0
+
+    ! equivDip
+    iCount = 0
+    do iAt = 1, nAtom
+      do ii = 1, dimDipole
+        iCount = iCount + 1
+        equivDip(ii, iAt) = iCount
+      end do
+    end do
+
+    ! equivQuad
+    iCount = 0
+    do iAt = 1, nAtom
+      do ii = 1, dimQuadrupole
+        iCount = iCount + 1
+        equivQuad(ii, iAt) = iCount
+      end do
+    end do
+
+  end subroutine getOrbitalEquiv
+
+
+  !> Initializes instance.
+  subroutine dftbMultiExpan_init(this, inp)
+
+    !> Instance.
+    type(TDftbMultiExpan), intent(out) :: this
+
+    !> Input data.
+    type(TDftbMultiExpanInp), intent(in) :: inp
+
+    integer, parameter :: xyzLen = 3
+    integer, parameter :: icx = 1, icy = 2, icz = 3
+    integer, parameter :: ios = 1, iopy = 2, iopz = 3, iopx = 4
+    integer, parameter :: iodxy = 5, iodyz = 6, iodzz = 7, iodxz = 8, iodxxyy = 9
+    real(dp), parameter :: tolZero = 1.0e-15_dp
+    integer :: nAtoms, maxNOrb, iAt1, iAt2, nOrb1, nOrb2, ii, jj, iSp1, iSp2
+    integer :: mu, nu, mm, nn
+    real(dp) tmpIntgrl, tmpTrace
+
+    this%nAtoms = size(inp%orb%nOrbAtom)
+    this%nSpecies = inp%nSpecies
+    allocate(this%nOrbSpecies(this%nSpecies))
+    this%nOrbSpecies(:) = inp%orb%nOrbSpecies(:)
+    this%mShells = 1
+    this%mShellsReal = inp%orb%mShell
+    this%nOrb = inp%nOrb
+    maxNOrb = maxval(inp%orb%nOrbSpecies(:))
+    this%nSpin = inp%nSpin
+    allocate(this%nShells(this%nSpecies))
+    this%nShells(:) = 1
+
+    allocate(this%hubbu(size(inp%hubbu)))
+    this%hubbu(:) = inp%hubbu
+
+    allocate(this%species(this%nAtoms))
+    this%species(:) = inp%species
+
+    allocate(this%onDQOCharges(3, this%nSpecies))
+    this%onDQOCharges(1,:) = 1
+    this%onDQOCharges(2,:) = 1
+    this%onDQOCharges(3,:) = 0
+
+    allocate(this%coords(xyzLen, this%nAtoms))
+    this%coords(:,:) = 0.0_dp
+
+    allocate(this%atomicDIntgrl(xyzLen, maxNOrb, maxNOrb, this%nSpecies))
+    allocate(this%atomicQIntgrl(xyzLen, xyzLen, maxNOrb, maxNOrb, this%nSpecies))
+    this%atomicDIntgrl(:,:,:,:) = 0.0_dp
+    this%atomicQIntgrl(:,:,:,:,:) = 0.0_dp
+
+    allocate(this%atomicOnsiteScaling(this%nSpecies))
+    this%atomicOnsiteScaling(:) = 1.0_dp
+
+
+    do iSp1 = 1, this%nSpecies
+      nOrb1 = this%nOrbSpecies(iSp1)
+
+      !Dipole
+      if ( nOrb1 >= 4 ) then
```

Comment:
The basis can be arbitrary, no algorithm should assume s, p, d ordering or the presence of all shells up to a given angular momentum.

Reply:
As mentioned, there are examples with basis sets of:
ss* test/app/dftb+/scc/HH_custom-ref/
ps test/app/dftb+/poisson/CH4-selShell/

Once it is initialised, orb%angShell(iShel, iSpecies) contains the angular momentum, counting from 0 of shell iShell of chemical species iSpecies.